app/uk/gov/hmrc/agentsubscriptionfrontend/controllers/BusinessIdentificationController.scala

```diff
@@ -265,25 +296,80 @@ class BusinessIdentificationController @Inject()(
     }
   }
 
-  private def lookupNextPage(knownFactsResult: KnownFactsResult)(implicit hc: HeaderCarrier) =
-    if (knownFactsResult.isSubscribedToAgentServices) {
-      mark("Count-Subscription-AlreadySubscribed-RegisteredInETMP")
-      Future.successful(routes.BusinessIdentificationController.showAlreadySubscribed())
-    } else {
-      sessionStoreService
-        .cacheInitialDetails(
-          InitialDetails(
-            knownFactsResult.utr,
-            knownFactsResult.postcode,
-            knownFactsResult.taxpayerName,
-            knownFactsResult.emailAddress,
-            knownFactsResult.address.getOrElse(throw new Exception("address object missing"))
-          ))
-        .map { _ =>
-          mark("Count-Subscription-CleanCreds-Start")
-          routes.SubscriptionController.showCheckAnswers()
+  private def lookupNextPage(initialDetails: InitialDetails)(implicit hc: HeaderCarrier) = {
+    val redirectCall = initialDetailsValidator.validate(initialDetails) match {
+      case Failure(responses) if responses.contains(InvalidBusinessName) =>
+        routes.BusinessIdentificationController.showBusinessNameForm()
+      case Failure(responses) if responses.contains(InvalidEmail) =>
+        routes.BusinessIdentificationController.showBusinessEmailForm()
+      case _ =>
+        mark("Count-Subscription-CleanCreds-Start")
+        routes.SubscriptionController.showCheckAnswers()
+    }
+
+    cacheInitialDetailsAndRedirect(initialDetails)(redirectCall)
+  }
+
+  val showBusinessEmailForm: Action[AnyContent] = Action.async { implicit request =>
+    withSubscribingAgent { _ =>
+      withInitialDetails { details =>
+        val form =
+          if (details.email.nonEmpty)
+            businessEmailForm.fill(BusinessEmail(details.email.get))
+          else businessEmailForm
+
+        Future.successful(Ok(html.business_email(form, details.email)))
+      }
+    }
+  }
+
+  val submitBusinessEmailForm: Action[AnyContent] = Action.async { implicit request =>
+    withSubscribingAgent { _ =>
+      withInitialDetails { details =>
+        businessEmailForm
+          .bindFromRequest()
+          .fold(
+            formWithErrors => Future.successful(Ok(html.business_email(formWithErrors, details.email))),
+            validForm => {
+              val updatedDetails = details.copy(email = Some(validForm.email))
+              sessionStoreService
+                .cacheInitialDetails(updatedDetails)
+                .flatMap(_ => lookupNextPage(updatedDetails).map(Redirect(_)))
+            }
+          )
+      }
+    }
+  }
+
+  val showBusinessNameForm: Action[AnyContent] = Action.async { implicit request =>
+    withSubscribingAgent { _ =>
+      withInitialDetails { details =>
+        val desErrorExists = initialDetailsValidator.validate(details) match {
```

Comment:
Do we need this `desErrorExists` knowlege also in the `submitBusinessNameForm` ?

Reply:
I /think/ it's ok. Seems the `submitBusinessNameForm` will always goes through `lookupNextPage()` on a validation error, which in turn always does a redirect to this endpoint.